### DISCUSSION TITLE
Implement Redis-compatible Protocol

### DISCUSSION THREAD

**Discussion by thuibr** [emmett-framework/granian#264] [Ideas]
I think it would be so interesting if Granian could also act as a Redis server. So much now depends on Redis being a part of the stack, from Django caching to Django Channels to Celery. I don't have time to take a stab at this, nor do I really know Rust, but Rust and Redis are interests of mine for future projects.

CodeCrafters has its [Redis project](https://app.codecrafters.io/courses/redis/overview) free for this month, which would be a good way to get up to speed. I also know that there are several Redis-compatible implementations out there. I wonder if a.) a Redis-compatible service could be built from the ground up in Granian, or b.) another implementation could be integrated into Granian.

I am just trying to start a discussion and see if there's any interest from the Granian maintainers and/or from somebody willing to take on this work.

Thanks!

**Comment by thuibr** [comment]
I suppose that the next logical step would be to also be able to run a celery worker. Maybe this is all out of scope for this project though. Feel free to close this if you think so. Thanks for all your work!

**Comment by gi0baro** [comment]
@thuibr I frankly think acting as a Redis instance would be definitely _out of scope_ for the project.
More over, due to Python GIL limitations, Granian relies quite a lot on multi-processing nowadays, thus implementing something like a memory storage would require *quite a lot of effort* into synchronising data across all the instances, and I don't really get the aim here. I might understand the hassle of managing Redis itself, but I still can't find a valid reason for an HTTP server to implement a memory storage solution.

What might happen in the future, given RSGI is designed primarily on Granian itself, is to introduce a _broadcasting feature_ to simplify the creations of _channels_ and share websockets, and eventually add this capability also to ASGI proposing an extension for it, but that would be _an alternative implementation of channels_, not a shared memory interface for its storage needs. And also, I'm actually talking about a _remote future_, this is definitely something I don't expect to see coming shortly.

**Reply by thuibr** [reply]
I think I sort of saw this project as an integration endeavor, consolidating the stack and making it easier to deploy fully-featured applications, but maybe that is entirely off the mark.
